### PR TITLE
TT-7306: Migrate Mock Response from Classic API Definition to OAS API Definition

### PR DESCRIPTION
**Pull request by edsonmichaque**
### **User description**
<details open>
  <summary><a href="https://tyktech.atlassian.net/browse/TT-7306" title="TT-7306" target="_blank">TT-7306</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
      <td>[OAS:migration] Migrate Mock Response from Classic API Definition to OAS API Definition</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Bug
        Bug
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Dev</td>
    </tr>
    <tr>
      <th>Points</th>
      <td>N/A</td>
    </tr>
    <tr>
      <th>Labels</th>
      <td>-</td>
    </tr>
  </table>
</details>

---

## Description
This PR adds support for converting `mock responses` between Tyk's classic API format and OAS format.

## Related Issue
https://tyktech.atlassian.net/browse/TT-7306

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added support for migrating mock responses between Classic and OAS API definitions.

- Implemented methods to fill and extract mock responses in OAS middleware.

- Enhanced test coverage for mock response handling, including edge cases.

- Introduced YAML fixtures for mock response scenarios in OAS and Classic formats.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>middleware.go</strong><dd><code>Enhance middleware to support mock responses</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/middleware.go

<li>Added initialization for <code>Operations</code> in the <code>Fill</code> method.<br> <li> Enhanced <code>Fill</code> method to handle mock responses.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6894/files#diff-992ec7c28d25fd54f6491d295389757705cd114bc869a35cba50d42e548cdc6e">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>operation.go</strong><dd><code>Add mock response handling in OAS operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/operation.go

<li>Added logic to handle mock response paths and operations.<br> <li> Implemented methods to fill and extract mock responses.<br> <li> Introduced <code>generateOperationID</code> utility for mock responses.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6894/files#diff-6d92d2d5b09a5fa7129609bb7cd0d383d015250ec07062b6a93a83257be51fb5">+139/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>middleware_test.go</strong><dd><code>Add tests for middleware mock response handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/middleware_test.go

<li>Added test case for mock response extraction and filling.<br> <li> Verified mock response consistency during migration.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6894/files#diff-0af31cb29ae298a6ac3e402b283ab364a6fd793fd04f253ef7c4983234c17bef">+26/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>operation_test.go</strong><dd><code>Add tests for OAS mock response operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/operation_test.go

<li>Added comprehensive tests for mock response extraction and filling.<br> <li> Covered scenarios for basic, multiple, and disabled mock responses.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6894/files#diff-cd234db716d6d2edc97c135ef546021c9ab4fa9282d63964bd155d41635cf964">+345/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>mock_response.yml</strong><dd><code>Add YAML fixtures for mock response testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/testdata/fixtures/mock_response.yml

<li>Added YAML fixtures for mock response scenarios.<br> <li> Included cases for OAS to Classic and vice versa.<br> <li> Covered enabled and disabled mock responses.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6894/files#diff-c7c72a9398d68abedf9238cc2a9606521069e13034f921e7a979d859e0559c8d">+121/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

**Comment by buger** [comment]
<p>This PR is too huge for one to review :broken_heart: </p>
  
    <table>
      <tr>
          <th>Additions</th>
          <td>1292 :no_good_woman: </td>
      </tr>
      <tr>
          <th>Expected</th>
          <td>:arrow_down: 800</td>
        </tr>
    </table>
    <p>
    Consider breaking it down into multiple small PRs.
    </p>
    <p>
      Check out this <a href="https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests">guide</a> to learn more about PR best-practices.
    </p>

**Comment by buger** [comment]
<p>
    Let's make that PR title a 💯 shall we? 💪
    </p>
    <p>
    Your <em>PR title</em> and <em>story title</em> look <strong>slightly different</strong>. Just checking in to know if it was intentional!
    </p>
    <table>
      <tr>
        <th>Story Title</th>
        <td>[OAS:migration] Migrate Mock Response from Classic API Definition to OAS API Definition</td>
      </tr>
      <tr>
          <th>PR Title</th>
          <td>wip: extract and fill mock responses from/to oas/classic</td>
        </tr>
    </table>
    <p>
      Check out this <a href="https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests">guide</a> to learn more about PR best-practices.
    </p>

File: apidef/oas/operation.go [hunk 1]
```diff
@@ -163,6 +165,83 @@ func (s *OAS) fillPathsAndOperations(ep apidef.ExtendedPathsSet) {
...
+		}
+
+		statusCode := strconv.Itoa(mockResponse.Code)
+
+		contentType := "text/plain"
+		if ct, ok := mockResponse.Headers["Content-Type"]; ok && ct != "" {
+			contentType = ct
+		} else if ct, ok := mockResponse.Headers["content-type"]; ok && ct != "" {
+			contentType = ct
+		}
```
Comment (titpetric) [inline apidef/oas/operation.go:191]: Bit of a code smell, headers have canonical formatting and we're checking two casings.
https://pkg.go.dev/net/http#CanonicalHeaderKey

File: apidef/oas/operation.go [hunk 2]
```diff
@@ -163,6 +165,83 @@ func (s *OAS) fillPathsAndOperations(ep apidef.ExtendedPathsSet) {
...
+
+		// Detect JSON content
+		if mockResponse.Body != "" {
+			var js interface{}
```
Comment (titpetric) [inline apidef/oas/operation.go:195]:
```suggestion
			var js interface{}
```

Could this be json.RawMessage?

Comment (edsonmichaque) [reply apidef/oas/operation.go:195]: Let fix that, thanks

File: apidef/oas/operation.go (hunk 1 above)
Comment (edsonmichaque) [reply apidef/oas/operation.go:191]: Thanks, let me correct it

Comment (edsonmichaque) [reply apidef/oas/operation.go]: Fixed it, PTAL

File: apidef/oas/operation.go (hunk 2 above)
Comment (edsonmichaque) [reply apidef/oas/operation.go]: Fixed it, PTAL

File: apidef/oas/operation.go [hunk 3]
```diff
@@ -163,6 +166,133 @@ func (s *OAS) fillPathsAndOperations(ep apidef.ExtendedPathsSet) {
...
+
+		response := &openapi3.Response{
+			Headers:     make(openapi3.Headers),
+			Description: &dummyDescription,
+		}
+
+		contentType := "text/plain"
+
+		for name, value := range mock.Headers {
+			if http.CanonicalHeaderKey(name) == "Content-Type" {
```
Comment (jeffy-mathew) [inline apidef/oas/operation.go:222]: nit
```suggestion
			if http.CanonicalHeaderKey(name) == header.ContentType {
```

File: apidef/oas/operation.go [hunk 4]
```diff
@@ -163,6 +166,133 @@ func (s *OAS) fillPathsAndOperations(ep apidef.ExtendedPathsSet) {
...
+				break
+			}
+		}
+
+		// We attempt to guess the content type by checking if the body is a valid JSON.
+		if mock.Body != "" {
+			var body json.RawMessage
+			if err := json.Unmarshal([]byte(mock.Body), &body); err == nil {
+				contentType = "application/json"
+			}
```
Comment (jeffy-mathew) [inline apidef/oas/operation.go:233]: this would also satisfy a simple string - [ref](https://go.dev/play/p/8BUaeNbdwIC)

How about we use a map to unmarshal? - [ref](https://go.dev/play/p/OrwjK7Rip7-)

Comment (edsonmichaque) [reply apidef/oas/operation.go:233]: You're right, thanks @jeffy-mathew.
A quick question, do you think we should also unmarshal to []interface{}, to cover arrays?

**Comment by edsonmichaque** [comment]
PTAL again @titpetric and @jeffy-mathew

File: apidef/oas/operation.go [hunk 5]
```diff
@@ -163,6 +166,130 @@ func (s *OAS) fillPathsAndOperations(ep apidef.ExtendedPathsSet) {
...
+// For each mock response, it:
+// 1. Creates an OAS operation with a unique ID (if it doesn't exist)
+// 2. Sets up the mock response with content type detection and example values
+// 3. Configures the operation to ignore authentication for this endpoint
+//
+// The content type is determined by:
+// - Checking the Content-Type header if present
+// - Attempting to parse the body as JSON
+// - Defaulting to text/plain if neither above applies
+func (s *OAS) fillMockResponsePaths(paths openapi3.Paths, ep apidef.ExtendedPathsSet) {
```
Comment (andrei-tyk) [inline apidef/oas/operation.go:186]: the function body is way too long, we should create helper functions whenever possible to increase readability and reduce complexity

File: apidef/oas/operation.go [hunk 6]
```diff
@@ -163,6 +166,130 @@ func (s *OAS) fillPathsAndOperations(ep apidef.ExtendedPathsSet) {
...
+			}
+		}
+
+		if operation.Responses == nil {
+			operation.Responses = openapi3.NewResponses()
+		}
+
+		// Response description is required by the OAS spec, but we don't have it in Tyk classic.
+		// So we're using a dummy value to satisfy the spec.
+		dummyDescription := "CHANGE ME"
```
Comment (andrei-tyk) [inline apidef/oas/operation.go:205]: "oasRequiredDummyValue" sounds more indicative than "CHANGE ME"

File: apidef/oas/operation.go [hunk 7]
```diff
@@ -163,6 +166,130 @@ func (s *OAS) fillPathsAndOperations(ep apidef.ExtendedPathsSet) {
...
+		response := &openapi3.Response{
+			Headers:     make(openapi3.Headers),
+			Description: &dummyDescription,
+		}
+
+		contentType := "text/plain"
+
+		// We attempt to guess the content type by checking if the body is a valid JSON.
+		if mock.Body != "" {
+			var jsonValue = []interface{}{
```
Comment (andrei-tyk) [inline apidef/oas/operation.go:216]: no need to create a slice of variants, its easier to read if you check for both

File: apidef/oas/operation.go [hunk 8]
```diff
@@ -163,6 +166,130 @@ func (s *OAS) fillPathsAndOperations(ep apidef.ExtendedPathsSet) {
...
+					break
+				}
+			}
+		}
+
+		// Check Content-Type header if present (takes precedence over auto-detection)
+		for name, value := range mock.Headers {
+			if http.CanonicalHeaderKey(name) == tykheader.ContentType {
+				contentType = value
+				break
```
Comment (andrei-tyk) [inline apidef/oas/operation.go:233]: if you split the method into multiple helpers. you will be able to just return from the underlying helper instead of breaking the for loop

File: apidef/oas/operation_test.go [hunk 9]
```diff
@@ -137,6 +142,293 @@ func TestOAS_PathsAndOperations(t *testing.T) {
...
+						},
+					},
+				},
+			},
+			want: func(t *testing.T, ep *apidef.ExtendedPathsSet) {
+				assert.Len(t, ep.WhiteList, 1)
+				mockResp := ep.WhiteList[0]
+				assert.Equal(t, "/test", mockResp.Path)
+				assert.Equal(t, "GET", mockResp.Method)
+				require.NotNil(t, mockResp.MethodActions["GET"])
```
Comment (andrei-tyk) [inline apidef/oas/operation_test.go:190]: use either assert or require

File: apidef/oas/operation_test.go [hunk 10]
```diff
@@ -277,3 +569,520 @@ func TestOAS_RegexPaths(t *testing.T) {
...
+					},
+				}},
+			},
+			want: func(t *testing.T, spec *OAS) {
+				assert.Len(t, spec.Paths, 1)
+				pathItem := spec.Paths["/test"]
+				require.NotNil(t, pathItem)
+
+				// Verify operation
+				require.NotNil(t, pathItem.Get)
```
Comment (andrei-tyk) [inline apidef/oas/operation_test.go:625]: use either only assert or only require

File: apidef/oas/operation.go (hunk 6 above)
Comment (edsonmichaque) [reply apidef/oas/operation.go]: Fixed, thank you. PTAL again

File: apidef/oas/operation.go (hunk 7 above)
Comment (edsonmichaque) [reply apidef/oas/operation.go]: Fixed, thank you. PTAL again

File: apidef/oas/operation.go (hunk 8 above)
Comment (edsonmichaque) [reply apidef/oas/operation.go]: Fixed, thank you. PTAL again

File: apidef/oas/operation.go [hunk 11]
```diff
@@ -197,11 +324,36 @@ func (s *OAS) extractPathsAndOperations(ep *apidef.ExtendedPathsSet) {
...
+					tykOp.extractMockResponsePaths(ep, path, method)
 					break
 				}
 			}
 		}
 	}
+
+	// Sort the mock response paths by path, method, and response code.
+	// This ensures a deterministic order of mock responses.
+	sort.Slice(ep.WhiteList, func(i, j int) bool {
```
[the remark this answers is not in the document]
Comment (edsonmichaque) [reply apidef/oas/operation.go]: Fixed, thank you. PTAL again

File: apidef/oas/operation_test.go (hunk 9 above)
Comment (edsonmichaque) [reply apidef/oas/operation_test.go:191]: Fixed, thank you. PTAL again

File: apidef/oas/operation_test.go (hunk 10 above)
Comment (edsonmichaque) [reply apidef/oas/operation_test.go:627]: Fixed, thank you. PTAL again

File: apidef/oas/operation_test.go [hunk 12]
```diff
@@ -277,3 +569,520 @@ func TestOAS_RegexPaths(t *testing.T) {
...
+					{Path: "/test", Method: "HEAD", Code: 200, Body: ``},
+					{Path: "/test", Method: "OPTIONS", Code: 200, Body: ``},
+				},
+			},
+			want: func(t *testing.T, spec *OAS) {
+				pathItem := spec.Paths["/test"]
+				require.NotNil(t, pathItem)
+
+				// Helper function to verify operation
+				verifyOperation := func(op *openapi3.Operation, method string, code int, body string) {
```
[the remark this answers is not in the document]
Comment (edsonmichaque) [reply apidef/oas/operation_test.go]: Fixed, thank you. PTAL again

**Comment by edsonmichaque** [comment]
Hi @andrei-tyk PTAL again

**Review by andrei-tyk — APPROVED** [review]

File: apidef/oas/operation_test.go [hunk 13]
```diff
@@ -137,6 +142,294 @@ func TestOAS_PathsAndOperations(t *testing.T) {
...
 }
 
+func TestOAS_MockResponse_extractPathsAndOperations(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name string
+		spec OAS
+		want func(t *testing.T, ep *apidef.ExtendedPathsSet)
+	}{
```
Comment (titpetric) [inline apidef/oas/operation_test.go:152]: type testCase struct {
}

tests := []testCase{
...

File: apidef/oas/operation.go [hunk 14]
```diff
@@ -866,3 +1048,57 @@ func (s *OAS) fillCircuitBreaker(metas []apidef.CircuitBreakerMeta) {
 		}
 	}
 }
+
+// detectMockResponseContentType determines the Content-Type of the mock response.
+// It first checks the headers for an explicit Content-Type, then attempts to detect
+// the type from the body content. Returns "text/plain" if no specific type can be determined.
+func detectMockResponseContentType(mock apidef.MockResponseMeta) string {
+	for name, value := range mock.Headers {
+		if http.CanonicalHeaderKey(name) == tykheader.ContentType {
```
Comment (titpetric) [inline apidef/oas/operation.go:1057]: make a local type for the const and don't import /header, not to extend usage

**Review by titpetric — APPROVED** [review]
minor feedback left, excellent effort on fixtures

File: apidef/oas/operation.go [hunk 15]
```diff
@@ -866,3 +1048,57 @@ func (s *OAS) fillCircuitBreaker(metas []apidef.CircuitBreakerMeta) {
...
+			return value
+		}
+	}
+
+	if mock.Body == "" {
+		return "text/plain"
+	}
+
+	// We attempt to guess the content type by checking if the body is a valid JSON.
+	var arrayValue = []json.RawMessage{}
```
Comment (jeffy-mathew) [inline apidef/oas/operation.go:1067]: ❤️ nice edge case

**Review by jeffy-mathew — APPROVED** [review]
